### PR TITLE
Dismiss approvals

### PR DESCRIPTION
**Pull request by samrabelachew**
Dismiss stale approvals when filtering out failed policy checks. We don't want to use approvals from policy owners that aren't for the latest PR commit.

High level pseudocode of policy check step:

- Build and run OPA policies via Conftest to identify if there are any policies that fail
- Filter out any of the failed policies by checking of GH approvals allow for a bypass:
    - Dismiss stale PR reviews (dismiss approvals from policy owners that aren’t after the latest commit)
    - Re-fetch a list of the latest approver usernames from GH
    - For each failed policy, identify if one of the approver usernames is also considered an owner of the policy
- Determine if policy check step succeeds based on if there are still any failing policies after the filter

Note: when is the policy check step run
- after plan generation is successful when a new commit is pushed
- after plan generation is successful when a user comments "atlantis plan"
- when a user submits a GH approval and atlantis finds that the current PR has failing policy checks

### CODE REVIEW

**Review by asubrocky — DISMISSED** [review]

File: server/events/policy_filter.go
Comment (nishkrishnan) [inline server/events/policy_filter.go:148]: it's not clear to me what the function does based on the naming, I'm assuming this checks whether any reviewers are policy owners.

Maybe `anyOwnerReviews` or something along those lines would make more sense.

File: server/events/policy_filter.go
Comment (nishkrishnan) [inline server/events/policy_filter.go:30]: remind me how this gets refreshed

File: server/events/policy_filter.go
Comment (samrabelachew) [reply server/events/policy_filter.go:106]: I would rather we stop the entire call instead of dismiss a review that may not have ever needed to be dismissed.

File: server/vcs/provider/github/pr_commit_fetcher.go
Comment (samrabelachew) [reply server/vcs/provider/github/pr_commit_fetcher.go:39]: Technically api returns pointers to these objects, so I figured with a pointer comes the risk of being nil. I don't expect them to be nil, but in case GH does have issues sending a full response, I think there is no harm in writing these nil checks to be safe.

File: server/events/policy_filter.go
Comment (samrabelachew) [reply server/events/policy_filter.go:29]: Any time there is a cache miss when looking for a specific user. https://github.com/lyft/atlantis/blob/d2345e7dfc389b0a5ab91537b37d63ad1a27bd55/server/events/policy_filter.go#L134-L145

Two cases we call for it:

- [Case 1](https://github.com/lyft/atlantis/blob/d2345e7dfc389b0a5ab91537b37d63ad1a27bd55/server/events/policy_filter.go#L76): When checking to see if the PR has an approval from a failed policy owner, thus we can filter out the failed policy
- [Case 2](https://github.com/lyft/atlantis/blob/d2345e7dfc389b0a5ab91537b37d63ad1a27bd55/server/events/policy_filter.go#L123): When first looking at all of the existing PR approvals to identify if any belong to an owner and can be dismissed

File: server/events/policy_filter.go
Comment (nishkrishnan) [reply server/events/policy_filter.go:105]: What's the UX if we do that?

File: server/events/policy_filter.go
Comment (nishkrishnan) [reply server/events/policy_filter.go:29]: > Any time there is a cache miss when looking for a specific user.

Is this right? What if someone gets removed from the owners list?  

If you're caching, typical practice is to to some time based expiry.  Managing deltas gets tricky and is somewhat error prone imo.

File: server/vcs/provider/github/pr_commit_fetcher.go
Comment (nishkrishnan) [reply server/vcs/provider/github/pr_commit_fetcher.go]: Ok, and is returning an error necessary here.  Again, would that impact the UX? Is there a failover mechanism we can apply here to minimize impact to the customer?

File: server/events/policy_filter.go
Comment (samrabelachew) [reply server/events/policy_filter.go:105]: For any error, the UX would see that the check run still fails. The original failed policy check result would be outputted to the check run details, but we would prepend a "internal server error" message to the top of it.

File: server/vcs/provider/github/pr_commit_fetcher.go
Comment (samrabelachew) [reply server/vcs/provider/github/pr_commit_fetcher.go]: For any error, the UX would see that the check run still fails. The original failed policy check result would be outputted to the check run details, but we would prepend a "internal server error" message to the top of it.

File: server/events/policy_filter.go
Comment (samrabelachew) [reply server/events/policy_filter.go]: Discussed offline, I removed cache entirely.

**Review by nishkrishnan — APPROVED** [review]

**Pull request merged**
